manager/manager.go

```diff
@@ -0,0 +1,228 @@
+					logger.Warn(
+						"Public IP's requested, but instance doesn't have a public IP! Skipping instance",
+						zap.String("instance-id", *instance.InstanceId),
+					)
+					continue
+				}
+				if m.config.RecordType == "A" {
+					targets = append(targets, *instance.PublicIpAddress)
+				}
+				if m.config.RecordType == "CNAME" {
+					targets = append(targets, *instance.PublicDnsName)
+				}
+			default:
+				if m.config.RecordType == "A" {
+					targets = append(targets, *instance.PrivateIpAddress)
+				}
+				if m.config.RecordType == "CNAME" {
+					targets = append(targets, *instance.PrivateDnsName)
+				}
+			}
+		}
+	}
+	return targets, err
+}
+
+// buildRoutes returns a map of the DNS names for the given IPs
+func (m Manager) buildRoutes(ips []string) map[string]string {
+	response := make(map[string]string)
+
+	sort.Strings(ips) // Sort ips for reproducible results
+
+	for i, ip := range ips {
+		response[fmt.Sprintf("%s%d.%s", m.config.SubdomainPrefix, i, m.Domain)] = ip
+	}
+	return response
+}
+
+// getRoute returns the value of a given DNS name. Returns NoRoute if
+// the route doesn't exist
+func (m Manager) getRoute(route string) (string, error) {
+	svc := route53.New(m.session)
+	params := &route53.ListResourceRecordSetsInput{
+		HostedZoneId:    aws.String(m.config.HostedZoneID), // Required
+		StartRecordName: aws.String(route),
+	}
+	resp, err := svc.ListResourceRecordSets(params)
+	if err != nil {
+		return "", err
+	}
+	if len(resp.ResourceRecordSets) == 0 {
+		return "", NoRecord
+	}
+	if len(resp.ResourceRecordSets[0].ResourceRecords) == 0 {
+		return "", NoRecord
+	}
+	return *resp.ResourceRecordSets[0].ResourceRecords[0].Value, nil
+}
+
+// defineUpdates creates a map of any route changes to be made
+func (m Manager) defineUpdates() (map[string]string, error) {
+	targets, err := m.getTargets()
+	if err != nil {
+		return nil, err
+	}
+	desiredRoutes := m.buildRoutes(targets)
+
+	updates := map[string]string{}
+
+	for k, v := range desiredRoutes {
+		existingRecord, err := m.getRoute(k)
+		if err != nil && err != NoRecord {
+			logger.Error(
+				"Couldn't determine the status of route",
+				zap.String("route", k),
+				zap.Error(err),
+			)
+			continue
+		}
+		if v != existingRecord {
+			updates[k] = v
+		}
+	}
+	return updates, nil
+}
+
+// assignRoutes performs the given route assignments
+func (m Manager) assignRoutes(assignments map[string]string) error {
+	svc := route53.New(m.session)
+
+	changes := []*route53.Change{}
+
+	for k, v := range assignments {
+		changes = append(changes, &route53.Change{
+			Action: aws.String("UPSERT"),
+			ResourceRecordSet: &route53.ResourceRecordSet{
+				Name:            aws.String(k),
+				Type:            aws.String(m.config.RecordType),
+				ResourceRecords: []*route53.ResourceRecord{{Value: aws.String(v)}},
+				TTL:             aws.Int64(m.config.TTL),
+			},
+		})
+	}
+
+	_, err := svc.ChangeResourceRecordSets(&route53.ChangeResourceRecordSetsInput{
```

Comment:
I'll be changing this to [ChangeResourceRecordSetsWithContext](http://docs.aws.amazon.com/sdk-for-go/api/service/route53/#Route53.ChangeResourceRecordSetsWithContext) so that the calling goroutine in `main.go` can give it a timeout context